src/c_api/c_api.cc

```diff
@@ -825,74 +825,67 @@ XGB_DLL int XGBoosterPredictFromDMatrix(BoosterHandle handle,
   API_END();
 }
 
-template <typename T>
-void InplacePredictImpl(std::shared_ptr<T> x, std::shared_ptr<DMatrix> p_m,
-                        char const *c_json_config, Learner *learner,
-                        size_t n_rows, size_t n_cols,
-                        xgboost::bst_ulong const **out_shape,
-                        xgboost::bst_ulong *out_dim, const float **out_result) {
+void InplacePredictImpl(std::shared_ptr<DMatrix> p_m, char const *c_json_config, Learner *learner,
+                        xgboost::bst_ulong const **out_shape, xgboost::bst_ulong *out_dim,
+                        const float **out_result) {
   auto config = Json::Load(StringView{c_json_config});
   CHECK_EQ(get<Integer const>(config["cache_id"]), 0) << "Cache ID is not supported yet";
 
-  HostDeviceVector<float>* p_predt { nullptr };
+  HostDeviceVector<float> *p_predt{nullptr};
   auto type = PredictionType(RequiredArg<Integer>(config, "type", __func__));
   float missing = GetMissing(config);
-  learner->InplacePredict(x, p_m, type, missing, &p_predt,
+  learner->InplacePredict(p_m, type, missing, &p_predt,
                           RequiredArg<Integer>(config, "iteration_begin", __func__),
                           RequiredArg<Integer>(config, "iteration_end", __func__));
   CHECK(p_predt);
   auto &shape = learner->GetThreadLocal().prediction_shape;
-  auto chunksize = n_rows == 0 ? 0 : p_predt->Size() / n_rows;
+  auto const &info = p_m->Info();
+  auto n_samples = info.num_row_;
+  auto n_features = info.num_col_;
+  auto chunksize = n_samples == 0 ? 0 : p_predt->Size() / n_samples;
   bool strict_shape = RequiredArg<Boolean>(config, "strict_shape", __func__);
-  CalcPredictShape(strict_shape, type, n_rows, n_cols, chunksize, learner->Groups(),
+  CalcPredictShape(strict_shape, type, n_samples, n_features, chunksize, learner->Groups(),
                    learner->BoostedRounds(), &shape, out_dim);
   *out_result = dmlc::BeginPtr(p_predt->HostVector());
   *out_shape = dmlc::BeginPtr(shape);
 }
 
 // A hidden API as cache id is not being supported yet.
```

Comment:
```suggestion
```
`XGBoosterPredictFromDense` is not a hidden API any more. It's listed in the C API header.